src/hotspot/share/opto/loopnode.hpp

```diff
@@ -1266,26 +1270,92 @@ class PhaseIdealLoop : public PhaseTransform {
-  Node* loop_exit_test(Node* back_control, IdealLoopTree* loop, Node*& incr, Node*& limit, BoolTest::mask& bt, float& cl_prob);
-  Node* loop_iv_incr(Node* incr, Node* x, IdealLoopTree* loop, Node*& phi_incr);
-  Node* loop_iv_stride(Node* incr, IdealLoopTree* loop, Node*& xphi);
-  PhiNode* loop_iv_phi(Node* xphi, Node* phi_incr, Node* x, IdealLoopTree* loop);
+  Node* loop_exit_control(const Node* x, const IdealLoopTree* loop) const;
+
+  struct LoopExitTest {
+    Node* cmp = nullptr;
+    Node* incr = nullptr;
+    Node* limit = nullptr;
+    const BoolTest::mask mask = BoolTest::illegal;
+    const float cl_prob = 0.0f;
+  };
+  LoopExitTest loop_exit_test(const Node* back_control, const IdealLoopTree* loop);
+
+  struct LoopIVIncr {
+    Node* incr = nullptr;
+    Node* phi_incr = nullptr;
+  };
+  LoopIVIncr loop_iv_incr(Node* old_incr, const Node* x, const IdealLoopTree* loop);
+
+  struct LoopIvStride {
+    Node* stride = nullptr;
+    Node* xphi = nullptr;
+  };
+  static LoopIvStride loop_iv_stride(const Node* incr);
 
-  bool is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_bt);
+  static PhiNode* loop_iv_phi(const Node* xphi, const Node* phi_incr, const Node* x);
+
+  bool try_convert_to_counted_loop(Node* head, IdealLoopTree*&loop, const BasicType iv_bt);
 
   Node* loop_nest_replace_iv(Node* iv_to_replace, Node* inner_iv, Node* outer_phi, Node* inner_head, BasicType bt);
   bool create_loop_nest(IdealLoopTree* loop, Node_List &old_new);
-#ifdef ASSERT
-  bool convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* loop);
-#endif
+
   void add_parse_predicate(Deoptimization::DeoptReason reason, Node* inner_head, IdealLoopTree* loop, SafePointNode* sfpt);
-  SafePointNode* find_safepoint(Node* back_control, Node* x, IdealLoopTree* loop);
+  SafePointNode* find_safepoint(Node* back_control, const Node* x, const IdealLoopTree* loop);
   IdealLoopTree* insert_outer_loop(IdealLoopTree* loop, LoopNode* outer_l, Node* outer_ift);
-  IdealLoopTree* create_outer_strip_mined_loop(BoolNode *test, Node *cmp, Node *init_control,
+  IdealLoopTree* create_outer_strip_mined_loop(Node *init_control,
                                                IdealLoopTree* loop, float cl_prob, float le_fcnt,
                                                Node*& entry_control, Node*& iffalse);
 
+  class CountedLoopConverter {
+    PhaseIdealLoop* _phase;
+    Node* _head;
+    IdealLoopTree* _loop;
+    const BasicType _iv_bt;
+
+    bool _insert_stride_overflow_limit_check = false;
+    bool _insert_init_trip_limit_check = false;
+
+#ifdef ASSERT
+    bool _checked_for_counted_loop = false;
+#endif
+
+    // To be assigned once a counted loop is confirmed.
+    Node* _limit;
+    jlong _stride_con;
+    Node* _phi;
+    Node* _phi_incr;
+    Node* _stride;
+    bool _includes_limit;
+    BoolTest::mask _mask;
+    Node* _increment;
+    Node* _cmp;
+    float _cl_prob;
+    Node* _sfpt;
+    jlong _final_correction;
+    Node* _trunc1;
```

Comment:
`trunc1` is a little hard to understand. Can we find a better name?

Reply:
Droped `Node* _trunc1` in favour of `TypeInteger* _increment_truncation_type`. Added a note in `should_stress_long_counted_loop()`.